### ISSUE TITLE
JRFC 27 - Hyper Modular Programming System

### ISSUE THREAD

**Issue by jbenet** [jbenet/random-ideas#27]
Over the last six months I've crossed that dark threshold where the desire of building a programming language has become an appealing idea. Terrifyingly, I _might_ actually build this some day. Well, not a language, a programming _system_. The arguments behind its design are long and will be written up some day, but for now I'll just dump the central tenet and core ideas here.
### An Illustration

You open your editor and begin to write a function. The first thing you do is write the ([mandatory](https://code.google.com/p/go-wiki/wiki/CodeReviewComments#Doc_Comments)) [doc comment](http://golang.org/doc/effective_go.html#commentary) describing what it does, and the type signature (yes, static typing). As you write, [your editor suggests](http://npmsearch.com/?q=factorial) lists of functions [published to the web](npmjs.org) (public or private) that match what you're typing. One appears promising, you inspect it. The editor loads the code. If it is exactly what you were going to write. You select it, and you're done.

If no result fits what you want, you continue to write the function implementation. You decompose the problem as much as possible, each time attempting to reuse existing functions. When done, you save it. The editor/compiler/system parses the text, analyzes + compresses the resulting ASG to try to find "the one way" (or good enough) to write the function. This representation is then content addressed, and a module consisting of the compresses representation, the source, and function metadata (doc string, author, version, etc) is published to the [(permanent) web](http://ipfs.io/), for everyone else to use.
### Important Ideas
- exporting a symbol (functions, classes, constants, ...) is the unit of modularity (node)
- system centered around writing _functions_ and writing them once (node)
- stress interfaces, decomposition, abstraction, types (haskell)
- use doc string + function signatures to suggest already published implementations (node, Go)
- content address functions based on compressed representations
- track version history of functions + dependents (in case there are bug fixes, etc). (node)
- if a function has a bug, can crawl the code importing it and notify dependents of bugfix. (node, Go)
- use static analysis to [infer version numbers](https://github.com/jbenet/random-ideas/issues/22): `<interface>.<implementation>` (semver inspired)
- when importing, you always bind to a version, but can choose to bind to `<interface>/<implementation>` or just `<interface>`
- e.g. `factorial = import QmZGhvJiYdp9Q/QmZGhvJiYdp9Q` (though editors can soften the ugly hashes) (node + ipfs)
- all modules (functions) are written and published to the (permanent) web (public or private)
- when importing a function, you import using its content address, and bind it to an explicit local name (`foo = import <function path>` type of thing) 
- the registry of all functions is [mounted locally](https://github.com/jbenet/random-ideas/issues/19) and accessible in the filesystem (ipfs style)
- _hyper modular_ means both to "very modular" and "modules are linked and on the web"

Note: this system is not about the language, it is about the machinery and process around producing, publishing, finding, reusing, running, testing, maintaining, auditing, bugfixing, republishing, and understanding code. (It's more about _the process of programming_, than _expressing programs_).  This means that the system only expresses constraints on language properties, and might work with modified versions of existing languages.

**Comment by vijayee** [comment]
So yeah this sounds to me very much like what Ethereum has going with smart-contracts but with package management going on. I wanted to build something similar to a smart-contract package manager for ethereum contracts but this sounds like a much better idea.

**Comment by vijayee** [comment]
https://www.nitrous.io/?utm_source=github.com&utm_campaign=Telescope&utm_medium=hackonnitrous

**Comment by adamdicarlo** [comment]
@vijayee uh... is that... spam? :frowning:

**Comment by jbenet** [comment]
@adamdicarlo no, i think it's somewhat relevant. look at the IDE shown there.

**Comment by vijayee** [comment]
@adamdicarlo I may not always spam github issues but when I do its relevant to distributed programming collaboration

**Comment by davidar** [comment]
I recall some [similar ideas](http://erlang.org/pipermail/erlang-questions/2011-May/058768.html) being discussed on the Erlang mailing list a while ago, which I quite liked. In terms of finding "the one way" to write a function, [Morte](http://www.haskellforall.com/2014/09/morte-intermediate-language-for-super.html) attempts to do this with (a subset of) Haskell. In fact, a surprisingly effective strategy for writing Haskell code is to first write the type signature for the function you want to write, then [Hoogle](https://www.haskell.org/hoogle/)/[Hayoo](http://hayoo.fh-wedel.de/) it to see if anyone has already written the corresponding definition. For some functions, it's even possible to [automatically generate](http://lambda-the-ultimate.org/node/1178) the definition given the type signature, by taking advantage of the Curry-Howard isomorphism and applying automated theorem-proving techniques. With ipfs/ipfs#4 and a suitable editor, it might even be possible to do a lot of this with unmodified Haskell.

**Comment by jbenet** [comment]
@davidar this would be great. Would love to hack together a prototype.

**Comment by davidar** [comment]
Also relevant: http://www.haskellforall.com/2015/05/the-internet-of-code.html

**Comment by mtwilliams** [comment]
Spit balling (seems to be the theme of this party):

Why not store edits on a AST? Imagine CQRS on the expression level. Sure, pragmatic reasons. (Imagine how unscalable that could be!) However you could do a bunch of ~~crazy~~ awesome stuff with it. For example, you could permute an application from a calendar to a game of pong. Or of more utility, do automatic bug fixes.

You could do similar things at the function level and take the Unix philosophy to the extreme.

**Comment by jbenet** [comment]
From https://github.com/ipfs/notes/issues/195

> Fantastic talk from the deeply thoughtful, idealistic and pragmatic, Rich Hickey: https://www.youtube.com/watch?v=oyLBGkS5ICk
>
> It captures much of what i dislike about semver. The thoughts arent completed though. Coupled with hashes, static analysis, and freedom from old paradigms (in their case, mvn, jars, namespaces, etc).
>
> It may be that Go's typing and packaging (however stunted...) is enough for something better along these lines. It may be that a new dep field for javascript could be enough. It may be that we need new languages whose import and code loading is structured around these _truly semantic_ versions.

**Comment by eparejatobes** [comment]
You might like https://github.com/unisonweb/unison/

**Comment by sehqlr** [comment]
This is something that I've also thought about, but from a different direction. I just wanted to add one piece of it for thought. Have a language ecosystem that has an semi-automated system to add to the standard library. The parts are:

- Language maintainers set up a specification to define requirements for submission, and a test suite that code MUST pass to be considered.
- if a lib passes the tests, then it can be added to the standard library. Whether or not it gets automatically approved or not is up to the language maintainers
- the stdlib site could also double as a social network/payments system for library authors as well

Am I going in a completely wrong direction? Are there parts of this idea that fits with what y'all are saying?

**Comment by sesam** [comment]
Neat! Software is resources (data, functions), state, storage, business logic and UX, and maybe that's it. Which of these are easier to handle as described above?
Easy early wins could be among: time, currencies, contacts (adress/phone/...), time-space-contexts, all full of exceptions and inconsistencies because life creates complexity.

Let's try imagining an example:
Say I want to build a new Android Home screen or Gear app with nearby people's free/busy schedules.
I could start defining:
1 calendar view of people's busy/free time
 1.1 data pagination
  1.1.1 get busy/free time data
   1.1.1.1 request calendar access
  1.1.2 merge busy/free time

Already typing the first line 1. maybe some or most of the points below 1. can be quick to just confirm to get included from a list of suggestions.

Reminds of TDD. Tests with coverage and verified correctness might come included with reusable packages, and can be written in any programming language. The module can run in a lambda or container on whatever hardware I have handy, maybe run it close to the data sources.

**Comment by omkarjc27** [comment]
I came saw this comment a couple of days ago.
I did something that is not as advanced as described above but ..... it is something.
So please Checkout [GitHub Link](https://github.com/omkarjc27/Snippet-Index)

**Comment by kehiy** [comment]
@jbenet That's a super cool idea which I always had similar ideas but on blockchain and smart contracts.

I saw this randomly and I think in these years with additional help of AI in coding, this kind of programming or web based (IPFS) libraries will be a new way to programming.

Here is my previous proposal specified for TON blockchain which is failed:
https://github.com/ton-blockchain/TEPs/pull/154

It had same way to import libraries (import "hash/content-address") and it also was trying to minimize them by merging same codes or even same part of some codes!!!

Also, based on IPFS paper, we can use it as a FS for VM. I think a VM can read some part of its byte code which is imported libraries from IPFS as well.

I will be happy to participate on this idea, a proposal for this or its implementations if any of them is in progress.
